File: Data Structures & Algorithms/lowest-common-ancestor-in-binary-search-tree/submission-0.js
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     constructor(val = 0, left = null, right = null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    /**
     * @param {TreeNode} root
     * @param {TreeNode} p
     * @param {TreeNode} q
     * @return {TreeNode}
     */
    lowestCommonAncestor(root, p, q) {
        function dfs(root, p, q){
            if(root==null){
               return null;
            }else{
                 if(root.val <p && root.val<q){
                   return dfs(root.right, p,q);
                }else if(root.val >p && root.val>q){
                    return dfs(root.left, p,q);
                }else{
                    return root;
                }
            }
        }

        return dfs(root, p.val, q.val);
    }
}

File: Data Structures & Algorithms/binary-tree-from-preorder-and-inorder-traversal/submission-0.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left),
 * right(right) {}
 * };
 */
class Solution {
public:
    TreeNode* buildTree(vector<int>& preorder, vector<int>& inorder) {
        int rootIdx = 0;
        return traverseAndBuildTree(preorder, inorder, 0, preorder.size() - 1,
                                    rootIdx);
    }

    TreeNode* traverseAndBuildTree(vector<int>& preorder, vector<int>& inorder,
                                   int startIdx, int endIdx, int& rootIdx) {
        if (startIdx > endIdx)
            return nullptr;
        TreeNode* root = new TreeNode(preorder[rootIdx]);
        int i = startIdx;
        for (; i < endIdx; i++) {
            if (inorder[i] == root->val) {
                break;
            }
        }
        rootIdx++;
        root->left =
            traverseAndBuildTree(preorder, inorder, startIdx, i - 1, rootIdx);
        root->right =
            traverseAndBuildTree(preorder, inorder, i + 1, endIdx, rootIdx);
        return root;
    }
};